tuplex/core/src/ee/local/Sort.cc

```diff
@@ -0,0 +1,605 @@
+    // the tuplex sort comparator is sorting (order is important, it should
+    // never change)
+    // const tuplex::Schema& schema: schema of the above partitions
+    // bool sortAsc: whethere to sort ascending or descending
+    // const std::vector<size_t>& indicesToSort: the columns to sort by
+    TuplexSortComparator(
+            std::vector<const uint8_t*> partitionPtrs,
+            const tuplex::Schema& schema, std::vector<size_t> order, std::vector<size_t> orderEnum) :
+            partitionPtrs(std::move(partitionPtrs)), schema(schema) {
+        int numColumns = schema.getRowType().parameters().size();
+        // generates the order of columns that the partition should be sorted by
+        this->order = genComparisonOrder(numColumns, order);
+        this->orderEnum = genComparisonOrderEnum(numColumns, orderEnum);
+    }
+
+    //    class SortBy(Enum):
+    //            ASCENDING = 1
+    //    DESCENDING = 2
+    //    ASCENDING_LENGTH = 3
+    //    DESCENDING_LENGTH = 4
+    //    ASCENDING_LEXICOGRAPHICALLY = 5
+    //    DESCENDING_LEXICOGRAPHICALLY = 6
+
+
+    TuplexSortComparator() = delete;
+
+    // returns true or false depending on which row should be sorted
+    bool operator() (PartitionSortType l, PartitionSortType r) const {
+        // gets the "left" row
+        tuplex::Row lRow = tuplex::Row::fromMemory(schema, partitionPtrs.at(l.first.first) + l.second.first,
+                                                   l.second.second - l.second.first);
+        // gets the "right" row
+        tuplex::Row rRow = tuplex::Row::fromMemory(schema, partitionPtrs.at(r.first.first) + r.second.first,
+                                                   r.second.second - r.second.first);
+        // gets a list of the types for each column
+        std::vector<python::Type> colTypes = schema.getRowType().parameters();
+        int counter = 0;
+        while (counter != order.size()) {
+            // gets the current column to sort by
+            int currentColIndex = order.at(counter);
+            // gets the current type of this current column
+            python::Type colType = colTypes.at(currentColIndex);
+            // sorting for integer
+            counter++;
+            if (colType == python::Type::I64) {
+                int64_t lInt = lRow.getInt(currentColIndex);
+                int64_t rInt = rRow.getInt(currentColIndex);
+                if (lInt == rInt) {continue;} // edge case where there is no other column to sort by
+                else if (orderEnum.at(counter-1) == 3) {
+                    std::ostringstream o1;
+                    std::ostringstream o2;
+                    o1 << lInt;
+                    o2 << rInt;
+                    std::string lStrInt = o1.str();
+                    std::string rStrInt = o2.str();
+                    return lStrInt.length() < rStrInt.length();
+                }
+                else if (orderEnum.at(counter-1) == 4) {
```

Comment:
maybe intro a helper variable storing result of counter - 1.

Reply:
also why does counter first get inced, and then when it's used always decreased by 1?